File: simulator/round_manager.py
from .action_resolver import *
from .actions.action_selector import get_action
from .misc import Statistics
from .resources import reset_resources
from .teams import Teams
from .utils.utils import preallocate_wildshape_forms
import logging
import copy

logger = logging.getLogger("Encounterra")


class RoundManager:
    def __init__(self, combatants, teams, effect_tracker, num_rounds=30):
        self.combatants = combatants
        self.teams = teams
        self.num_rounds = num_rounds
        self.effect_tracker = effect_tracker
        self.action_resolver = ActionResolver(combatants, teams, self.effect_tracker)

    def roll_initiative(self):
        for combatant in self.combatants:
            combatant.roll_initiative()

    def order_by_initiative(self):
        def by_initiative(e):
            return e.curr_init

        self.combatants.sort(key=by_initiative, reverse=True)
        logger.info("--------------INITIATIVE ORDER--------------")
        for combatant in self.combatants:
            logger.info(f"{combatant} with {combatant.curr_init}", extra={"team": combatant.team_color})

    def prep_combatants(self):
        """
        Secondary initialization. It's an optimization measure.
        """
        for combatant in self.combatants:
            for baf in combatant.bonus_action_factories:
                if baf[0] is BonusAction.MOON_WILDSHAPE:
                    combatant.available_wildshape_forms = preallocate_wildshape_forms(combatant, BonusAction.MOON_WILDSHAPE, baf[1])
                    break
            for af in combatant.action_factories:
                if af[0] is Action.WILDSHAPE:
                    combatant.available_wildshape_forms = preallocate_wildshape_forms(combatant, Action.WILDSHAPE, af[1])
                    break


    def goes_before_in_initiative(self, combatant1, combatant2):
        return True if self.combatants.index(combatant1) < self.combatants.index(combatant2) else False

    def is_only_one_team_standing(self):
        return True if len(self.teams.get_surviving_teams()) == 1 else False

    def reset(self, combatant_initial_positions):
        for combatant in self.combatants:
            reset_resources(combatant)
        self.effect_tracker.reset()
        Map.get().reset(combatant_initial_positions)


    def simulate_n(self, n=1, result_queue=None):
        if n > 0:
            team_tally = {color: {stat: 0 for stat in Statistics} for color in self.teams.get_team_colors()}
            combatant_initial_positions = {c: copy.deepcopy(Map.get().get_combatant_position(c).get()[0]) for c in self.combatants}
            self.prep_combatants()
            for i in range(n):
                logger.warning(f"{i}. Iteration")
                self.simulate()
                surviving_teams = self.teams.get_surviving_teams()
                if len(surviving_teams) > 1:
                    logger.warning("There's more than one surviving team. Battle's not over yet!")
                elif len(surviving_teams) == 0:
                    logger.warning("Everyone's dead. No winners!")
                else:
                    logger.warning(f"Team {surviving_teams[0].name} wins")
                    team_tally[surviving_teams[0]][Statistics.VICTORIES] += 1
                    dead_blue, dead_red = self.teams.get_death_count()
                    if dead_blue > 0:
                        team_tally[Teams.Color.BLUE][Statistics.AT_LEAST_ONE_DIED] += 1
                    if dead_red > 0:
                        team_tally[Teams.Color.RED][Statistics.AT_LEAST_ONE_DIED] += 1
                    if dead_blue > 1:
                        team_tally[Teams.Color.BLUE][Statistics.AT_LEAST_TWO_DIED] += 1
                    if dead_red > 1:
                        team_tally[Teams.Color.RED][Statistics.AT_LEAST_TWO_DIED] += 1
                    if dead_blue > 2:
                        team_tally[Teams.Color.BLUE][Statistics.AT_LEAST_THREE_DIED] += 1
                    if dead_red > 2:
                        team_tally[Teams.Color.RED][Statistics.AT_LEAST_THREE_DIED] += 1

                self.reset(combatant_initial_positions)
            if result_queue:
                result_queue.put(team_tally)
            # logger.warning("--------------STATISTICS--------------")
            # for name, victories in team_tally.items():
            #     logger.warning(f"Team {name.name} won total of {victories} times", extra={"team": name})
            return team_tally
        else:
            logger.error("Wrong input. n has to be 1 or higher!")
        return None

    def simulate(self):
        self.roll_initiative()
        self.order_by_initiative()
        done = False
        logger.info("--------------START--------------")
        for r in range(self.num_rounds):
            logger.info(f"Round {r + 1}:")
            if done:
                logger.info("The fight is over")
                break
            for combatant in self.combatants:
                if done:
                    break
                if not combatant.is_alive():
                    continue
                logger.info(f"It's {combatant}'s turn")
                logger.info(Map.get())
                self.effect_tracker.start_of_turn(combatant)
                if not combatant.is_alive():  # Start of turn effects can kill
                    continue
                combatant.new_turn()
                effects = self.effect_tracker.get_affecting_combatant(combatant)  # TODO consider cleaning this up by merging with start_of_turn
                self.action_resolver.resolve_effects(effects, combatant)
                if combatant.is_affected_by_any(Conditions.INCAPACITATED, Conditions.STUNNED, Conditions.PARALYZED, Conditions.PETRIFIED, Conditions.UNCONSCIOUS):
                    logger.info(f"{combatant} is affected by a condition which prevents any action. Skipping turn")
                    # self.effect_tracker.start_of_turn(combatant)
                    self.effect_tracker.end_of_turn(combatant)
                    combatant.on_end_of_turn()
                    continue
                while True:
                    action = get_action(combatant)
                    if action is None:
                        break
                    resolution = self.action_resolver.resolve_action(action, combatant)
                    if resolution is None:
                        break
                    if self.is_only_one_team_standing():
                        done = True
                        break
                    if not combatant.is_alive():
                        self.effect_tracker.combatant_died(combatant)
                        break  # could have died as a result of AoO
                if combatant.is_alive():
                    self.effect_tracker.end_of_turn(combatant)
                    combatant.on_end_of_turn()
            logger.info("----------------------------------")
            self.print_status()

    def print_status(self):
        for combatant in self.combatants:
            combatant = combatant.get_current_form()
            status = f"alive with {combatant.curr_hp} hp" if combatant.is_alive() else "dead"
            logger.info(f"{combatant} is {status}", extra={"team": self.teams.get_team(combatant)})
        logger.info(Map.get())

    def print_results(self):
        logger.info("--------------RESULT--------------")
        self.print_status()

File: c++/core/round_manager.cpp
#include "core/round_manager.hpp"
#include "actions/action_selection.hpp"
#include "abilities/wildshape_utils.hpp"

namespace enc
{

  void RoundManager::rollInitiative()
  {
    for(auto &combatant : _combatants)
      {
        combatant->rollInitiative();
      }
  }

  void RoundManager::orderByInitiative()
  {
    std::sort(_combatants.begin(), _combatants.end(),
              [](const std::shared_ptr<Combatant> &a, const std::shared_ptr<Combatant> &b) { return a->getCurrentInit() > b->getCurrentInit(); });

    std::cout << "--------------INITIATIVE ORDER--------------\n";
    for(const auto &combatant : _combatants)
      {
        std::cout << combatant->toString() << " with " << combatant->getCurrentInit() << "\n";
      }
  }

  void RoundManager::prepCombatants()
  {
    for(auto &combatant : _combatants)
      {
        // Check for moon & regular wildshape
        for(const auto &factory : combatant->getBonusActionFactoriesConst())
          {
            if(factory->getAbilityType() == AbilityType::MOON_WILDSHAPE || factory->getAbilityType() == AbilityType::WILDSHAPE)
              {
                auto *wildshapeFactory = static_cast<WildshapeFactory *>(factory.get());
                combatant->setAvailableWildshapeForms(preallocateWildshapeForms(combatant, factory->getAbilityType(), *wildshapeFactory));
                break;
              }
          }
      }
  }

  bool RoundManager::goesBeforeInInitiative(const Combatant &combatant1, const Combatant &combatant2) const
  {
    auto it1 = std::find(_combatants.begin(), _combatants.end(), combatant1);
    auto it2 = std::find(_combatants.begin(), _combatants.end(), combatant2);
    return it1 < it2;
  }

  bool RoundManager::isOnlyOneTeamStanding() const { return Teams::getInstance().getSurvivingTeams().size() == 1; }

  void RoundManager::reset(const std::unordered_map<int, Coords> &combatantInitialPositions)
  {
    EffectTracker::getInstance().reset();
    for(auto &combatant : _combatants)
      {
        combatant->reset();
      }
    BattleMap::getInstance().resetCombatantsToInitialPositions(combatantInitialPositions);
  }

  std::unordered_map<Color, std::unordered_map<Statistics, int>>
  RoundManager::simulateN(int n, std::queue<std::unordered_map<Color, std::unordered_map<Statistics, int>>> *resultQueue)
  {
    if(n <= 0)
      {
        std::cout << "Wrong input. n has to be 1 or higher!\n";
        return {};
      }

    std::unordered_map<Color, std::unordered_map<Statistics, int>> teamTally;
    for(Color color : Teams::getInstance().getTeamColors())
      {
        teamTally[color] = std::unordered_map<Statistics, int>();
        for(const auto &stat : getAllStatistics())
          {
            teamTally[color][stat] = 0;
          }
      }

    auto &battleMap = BattleMap::getInstance();
    std::unordered_map<int, Coords> combatantInitialPositions;
    for(const auto &combatant : _combatants)
      {
        combatantInitialPositions.emplace(combatant->_instanceId, battleMap.getCombatantCoordinates(*combatant));
      }

    prepCombatants();

    for(int i = 0; i < n; ++i)
      {
        std::cout << i << ". Iteration\n";
        simulate();

        auto survivingTeams = Teams::getInstance().getSurvivingTeams();
        if(survivingTeams.size() > 1)
          {
            std::cout << "There's more than one surviving team. Battle's not over yet!\n";
          }
        else if(survivingTeams.empty())
          {
            std::cout << "Everyone's dead. No winners!\n";
          }
        else
          {
            std::cout << "Team " << COLOR_NAMES.at(survivingTeams[0]) << " wins\n";
            teamTally[survivingTeams[0]][Statistics::VICTORIES]++;

            auto [deadBlue, deadRed] = Teams::getInstance().getDeathCount();
            if(deadBlue > 0)
              teamTally[Color::BLUE][Statistics::AT_LEAST_ONE_DIED]++;
            if(deadRed > 0)
              teamTally[Color::RED][Statistics::AT_LEAST_ONE_DIED]++;
            if(deadBlue > 1)
              teamTally[Color::BLUE][Statistics::AT_LEAST_TWO_DIED]++;
            if(deadRed > 1)
              teamTally[Color::RED][Statistics::AT_LEAST_TWO_DIED]++;
            if(deadBlue > 2)
              teamTally[Color::BLUE][Statistics::AT_LEAST_THREE_DIED]++;
            if(deadRed > 2)
              teamTally[Color::RED][Statistics::AT_LEAST_THREE_DIED]++;
          }

        reset(combatantInitialPositions);
      }

    if(resultQueue)
      {
        resultQueue->push(teamTally);
      }

    return teamTally;
  }

  void RoundManager::simulate()
  {
    auto &battleMap = BattleMap::getInstance();
    auto &effectTracker = EffectTracker::getInstance();

    rollInitiative();
    orderByInitiative();
    bool done = false;

    std::cout << "--------------START--------------\n";
    for(int r = 0; r < _numRounds; ++r)
      {
        battleMap.setCombatRound(r);
        std::cout << "Round " << (r + 1) << ":\n";

        if(done)
          {
            std::cout << "The fight is over\n";
            break;
          }

        for(auto &combatant : _combatants)
          {
            if(done)
              {
                break;
              }
            if(!combatant->isAlive())
              {
                continue;
              }

            std::cout << "It's " << combatant->toString() << "'s turn\n";
            _currCombatant = combatant;
            std::cout << battleMap.toString() << "\n";

            combatant->rollForRecharge();
            effectTracker.startOfTurnTick(*combatant);
            effectTracker.startOfTurn(*combatant);

            if(!combatant->isAlive())
              {
                continue; // Start of turn effects can kill
              }

            combatant->newTurn();
            auto effects = effectTracker.getAffectingCombatant(*combatant);
            resolveEffects(effects, *combatant);

            if(combatant->isAffectedByAny(
                 {Conditions::INCAPACITATED, Conditions::STUNNED, Conditions::PARALYZED, Conditions::PETRIFIED, Conditions::UNCONSCIOUS}))
              {
                std::cout << combatant->toString() << " is affected by a condition which prevents any action. Skipping turn\n";
                effectTracker.endOfTurn(*combatant);
                combatant->onEndOfTurn();
                continue;
              }

            if(isOnlyOneTeamStanding())
              {
                done = true;
                break;
              }

            while(true)
              {
                auto action = getAction(*combatant);
                if(!action)
                  {
                    break;
                  }

                ActionResult resolution = resolveAction(action, *combatant);
                if(resolution == ActionResult::UNFEASIBLE)
                  {
                    break;
                  }

                if(isOnlyOneTeamStanding())
                  {
                    done = true;
                    break;
                  }

                if(!combatant->isAlive())
                  {
                    effectTracker.combatantDied(*combatant);
                    break;
                  }
              }

            if(combatant->isAlive())
              {
                effectTracker.endOfTurn(*combatant);
                combatant->onEndOfTurn();
              }
          }

        std::cout << "----------------------------------\n";
        printStatus();
      }
  }

  void RoundManager::printStatus()
  {
    for(const auto &combatant : _combatants)
      {
        Combatant &currentForm = combatant->getCurrentForm();
        std::string status = currentForm.isAlive() ? "alive with " + std::to_string(currentForm.getCurrentHp()) + " hp" : "dead";

        std::cout << combatant->toString() << " is " << status << "\n";
      }
  }

  void RoundManager::printResults()
  {
    std::cout << "--------------RESULT--------------\n";
    printStatus();
  }

} // namespace enc
